### ISSUE TITLE
Change in element type of `setdiff` output from Julia 1.7 to 1.8

### ISSUE THREAD

**Issue by mtfishman** [JuliaLang/julia#46657]
In Julia 1.7 I see:
```julia
julia> setdiff([1, 2, 3], [1.0, 2.0])
1-element Vector{Int64}:
 3

julia> versioninfo()
Julia Version 1.7.3
Commit 742b9abb4d (2022-05-06 12:58 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Xeon(R) E-2176M  CPU @ 2.70GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-12.0.1 (ORCJIT, skylake)
Environment:
  JULIA_EDITOR = vim
```
while in Julia 1.8 I see:
```julia
julia> setdiff([1, 2, 3], [1.0, 2.0])
1-element Vector{Float64}:
 3.0

julia> versioninfo()
Julia Version 1.8.0
Commit 5544a0fab76 (2022-08-17 13:38 UTC)
Platform Info:
  OS: Linux (x86_64-linux-gnu)
  CPU: 6 × Intel(R) Xeon(R) E-2176M  CPU @ 2.70GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-13.0.1 (ORCJIT, skylake)
  Threads: 1 on 6 virtual cores
Environment:
  LD_PRELOAD = libgtk3-nocsd.so.0
  LD_LIBRARY_PATH = /usr/lib/x86_64-linux-gnu/libcutensor/11.1/:/opt/intel/compilers_and_libraries_2020.0.166/linux/tbb/lib/intel64_lin/gcc4.7:/opt/intel/compilers_and_libraries_2020.0.166/linux/compiler/lib/intel64_lin:/opt/intel/compilers_and_libraries_2020.0.166/linux/mkl/lib/intel64_lin::/home/mfishman/software/itensor/lib
  JULIA_EDITOR = vim
```
This broke some code in ITensors.jl which implicitly assumed the previous behavior. It's not hard to work around this and I'm not sure if there was an implied guarantee that this wouldn't change from version to version, I'm curious to hear if this was intended/what led to this change.

The behavior of 1.7 makes more sense to me since `setdiff` outputs the elements unique to the first input, so it seems like it should have the same element type as the first input.

**Comment by KristofferC** [comment]
https://github.com/JuliaLang/julia/issues/45511

**Comment by Seelengrab** [comment]
I don't have 1.7 installed, but I'm curious what you get for this:

```
julia> setdiff([1,2,3], Float64[1,2,3.1], ['a', 'b', 'c'])
1-element Vector{Any}:
 3
```

**Comment by mtfishman** [comment]
On Julia 1.7 I get:
```julia
julia> setdiff([1,2,3], Float64[1,2,3.1], ['a', 'b', 'c'])
1-element Vector{Int64}:
 3
```
and on Julia 1.8 I get:
```julia
julia> setdiff([1,2,3], Float64[1,2,3.1], ['a', 'b', 'c'])
1-element Vector{Any}:
 3
```

**Comment by rfourquet** [comment]
This was changed in #41769. Fwiw, I think this change was a mistake (https://github.com/JuliaLang/julia/pull/41769#issuecomment-898436916).

**Comment by adienes** [comment]
I think this is a regression and should be reverted. unlike the `union` docstring, which says ` Construct an object containing all distinct elements from all of the arguments.`

`setdiff` claims to contain elements specifically from the first collection which leads to annoying problems like

```
julia> setdiff([typemax(Int), 2, 3], [1.0, 2.0])
ERROR: TypeError: 9223372036854775807 is not a valid key for type Float64
```

**Comment by Seelengrab** [comment]
> setdiff claims to contain elements specifically from the first collection which leads to annoying problems like

Oof, that's a tricky one. This is the docstring:

> Construct the set of elements in s but not in any of the iterables in itrs. Maintain order with arrays.

Now, according to what measure is that "in" meant? Is it hash equality, `==` or `===`? Depending on that, returning `1.0` is valid behavior!

I think it makes sense to enforce that the `eltype` of the returned collection is the same as that of the first argument though. So even with hash equality, that additional constraint fixes this behavior.

**Comment by adienes** [comment]
IMO it should use the `in` function on the trailing arguments. i.e. should be basically equivalent to
```
function setdiff(x, args...)
    newx = emptymutable(x)
    for v in x
        any(v in a for a in args) && continue
        push!(newx, v)
    end
    newx
end
```

**Comment by Seelengrab** [comment]
Right, this version enforces the same `eltype` semantic. Just also consider:

```
julia> 1.0 in [1,2,3]
true
```

What I meant above with equality is exactly this, since saying that the element `1.0` is `in` the first argument `[1,2,3]` is the same as saying `1 in [1.0, 2.0, 3.0]`. So only looking at `in` is not enough IMO, because of that symmetry.

**Comment by adienes** [comment]
sorry, I'm still not sure I follow., since `1 in [1.0] == true` still. are you suggesting there could be some `A, y in B` such that `y in A` and `y == x` for some `x in A` but `!(x in B)` ?

would we agree on the following invariants?
```
S = setdiff(A, B)

@test eltype(S) == eltype(A)
@test !any(x in B for x in A)
```

**Comment by Seelengrab** [comment]
> sorry, I'm still not sure I follow., since 1 in [1.0] == true still. are you suggesting there could be some A, y in B such that y in A and y == x for some x in A but !(x in B)?

No, I'm not talking about any of the other arguments. I'm only talking about a possible justification for why `setdiff([1,2,3], [2.0, 3.0]) == [1.0]`, since `1.0 in [1,2,3]` is true due to how `in` works/which kind of equality `in` uses. The current docstring of `setdiff` doesn't specify that, so in that sense getting `[1.0]` is valid. The way to fix this is to either specify which equality must be used (which would influence the return values as well! E.g. choosing `===` would make `setdiff([1,2,3], [2.0, 3.0]) ` return `[1,2,3]`, since `1 !== 1.0`), or to enforce that the eltype of the returned collection matches the eltype of the first collection (which preserves the return value, just changes its type back to what it used to be).

> would we agree on the following invariants?

Yes, this is what I proposed above with the eltype invariant. The other one is already in the docstring.

**Comment by adienes** [comment]
ok I think I understand. given the distinction

* "in" as "resulting from iteration of" 
* "in" as the actual function `in`

I was implicitly reading the docstring as

> Construct the set of elements [resulting from iteration of] s but not [function in] any of the iterables

and you are saying, using the second definition consistently for both occurrences of the word, that `[1.0]` satisfies

> Construct the set of elements [function in] s but not [function in] any of the iterables

which is probably better, because my original interpretation cannot handle `setdiff` of objects that may not be finitely iterable, such as continuous ranges

but I do think there is still a key term in the docstring to highlight which is "the" set which implies uniqueness. since there could any number of `[MyOne]` with `MyOne == 1`we have to choose a canonical representative. this I think is a pretty strong argument to choose the one most identical to the input, including up to object identity in the case of mutable elements.

**Comment by LilithHafner** [comment]
Triage thinks we should restore the 1.7 behavior and clarify the docstring.
